src/main/engine/services/playwright.ts

```diff
@@ -0,0 +1,124 @@
+export class PlaywrightService {
+    protected browser?: Browser;
+    protected context?: BrowserContext;
+    protected targetId?: string;
+    protected currentPage?: Page;
+    protected cachedPages = new Map<string, Page>();
+    protected endpointUrl: string;
+
+    constructor(
+        @inject(Logger)
+        protected logger: Logger,
+        @inject(Configuration)
+        protected config: Configuration,
+    ) {
+        this.endpointUrl = `http://${this.config.get(CHROME_ADDRESS)}:${this.config.get(CHROME_PORT)}`;
+    }
+
+    async connectOverCDP(options?: ConnectOverCDPOptions) {
+        if (this.browser) {
+            await this.browser.close();
+        }
+
+        this.browser = await chromium.connectOverCDP(this.endpointUrl, options);
+        this.browser.on('disconnected', async () => {
+            await this.disconnect();
+        });
+        if (this.browser.contexts().length > 0) {
+            this.context = this.browser.contexts()[0];
+            this.context.on('page', async page => {
+                const pageTargetId = await this.getPageTargetId(page);
+                if (pageTargetId) {
+                    this.cachedPages.set(pageTargetId, page);
+                    page.on('close', async page => {
+                        for (const [key, value] of this.cachedPages) {
+                            if (value === page) {
+                                this.cachedPages.delete(key);
+                            }
+                        }
+                    });
+                }
+            });
+
+            await this.fillCacheForAllPages(this.context.pages());
+            if (!this.targetId) {
+                this.currentPage = this.context.pages()[0];
+            } else {
+                this.currentPage = this.cachedPages.get(this.targetId);
+            }
+        }
+    }
+
+    async disconnect() {
+        await this.browser?.close();
+        this.browser = undefined;
+        this.context = undefined;
+        this.currentPage = undefined;
+        this.targetId = undefined;
+        this.cachedPages.clear();
+    }
+
+    protected async fillCacheForAllPages(pages: Page[]) {
+        for (const page of pages) {
+            const pageTargetId = await this.getPageTargetId(page);
+            if (pageTargetId) {
+                this.cachedPages.set(pageTargetId, page);
+            }
+        }
+    }
+
+    async setCurrentPage(targetId: string) {
+        this.currentPage = undefined;
+        this.targetId = targetId;
+        if (!this.browser) {
+            await this.connectOverCDP();
+
+            if (!this.context) {
+                this.logger.warn(`Browser not running, failed to attach to ${targetId}`);
+                return;
+            }
+        }
+
+        if (this.cachedPages.has(targetId)) {
+            this.currentPage = this.cachedPages.get(targetId);
+            return;
+        }
+
+        const reversedPages = this.context!.pages().slice().reverse();
```

Comment:
Maybe [this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/toReversed) will make it more obvious what we are doing

```suggestion
        const reversedPages = this.context!.pages().toReversed();
```

Reply:
Also, is there any other reason to go through the reverse list other than to try to find the `targetId` faster (assuming it was one of the latest to be added)?

Reply:
Initially I wanted to use `toReversed()` but it got available since ES2023 so I didn't want to bump Typescript to version >5.2 just for that. 

Just to find it faster. It's actually to cover an edge case when new target is created and immediately attached. In that case this happens before [on page event is triggered](https://github.com/ubio/engine/blob/ff841916f1343212195af18a743a0312fb7d0bb0/src/main/engine/services/playwright.ts#L37) which adds page to cache.